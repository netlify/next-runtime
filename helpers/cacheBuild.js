const path = require('path')

const DEFAULT_DIST_DIR = '.next'

// Account for possible custom distDir
const getPath = (nextRoot, distDir, source) => path.join(nextRoot, distDir || DEFAULT_DIST_DIR, source)

const restoreCache = async ({ cache, distDir, nextRoot }) => {
  const cacheDir = getPath(nextRoot, distDir, 'cache')
  if (await cache.restore(cacheDir)) {
    console.log('Next.js cache restored.')
  } else {
    console.log('No Next.js cache to restore.')
  }
}

const saveCache = async ({ cache, distDir, nextRoot }) => {
  const cacheDir = getPath(nextRoot, distDir, 'cache')

  const buildManifest = getPath(nextRoot, distDir, 'build-manifest.json')
  if (await cache.save(cacheDir, { digests: [buildManifest] })) {
    console.log('Next.js cache saved.')
  } else {
    console.log('No Next.js cache to save.')
  }
}

module.exports = {
  restoreCache,
  saveCache,
}
