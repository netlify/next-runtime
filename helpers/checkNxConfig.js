const { existsSync } = require('fs')
const { EOL } = require('os')
const path = require('path')

const checkNxConfig = ({ netlifyConfig, nextConfig, failBuild, constants: { PUBLISH_DIR = 'out' } }) => {
  const errors = []
  if (nextConfig.distDir === '.next') {
    errors.push(
      "- When using Nx you must set a value for 'distDir' in your next.config.js, and the value cannot be '.next'",
    )
  }
  // The PUBLISH_DIR constant is normalized, so no leading slash is needed
  if (!PUBLISH_DIR.startsWith('apps/')) {
    errors.push(
      "Please set the 'publish' value in your Netlify build config to a folder inside your app directory. e.g. 'apps/myapp/out'",
    )
  }

  // Look for the config file as a sibling of the publish dir
  const expectedConfigFile = path.resolve(netlifyConfig.build.publish || PUBLISH_DIR, '..', 'next.config.js')

  if (expectedConfigFile !== nextConfig.configFile) {
    const confName = path.relative(process.cwd(), nextConfig.configFile)
    errors.push(
      `- Using incorrect config file '${confName}'. Expected to use '${path.relative(
        process.cwd(),
        expectedConfigFile,
      )}'`,
    )

    if (existsSync(expectedConfigFile)) {
      errors.push(
        `Please move or delete '${confName}'${confName === 'next.config.js' ? ' from the root of your site' : ''}.`,
      )
    } else {
      errors.push(
        `Please move or delete '${confName}'${
          confName === 'next.config.js' ? ' from the root of your site' : ''
        }, and create '${path.relative(process.cwd(), expectedConfigFile)}' instead.`,
      )
    }
  }

  if (errors.length !== 0) {
    failBuild(
      // TODO: Add ntl.fyi link to docs
      [
        'Invalid configuration',
        ...errors,
        'See the docs on using Nx with Netlify for more information: https://ntl.fyi/nx-next',
      ].join(EOL),
    )
  }
}

module.exports = checkNxConfig
