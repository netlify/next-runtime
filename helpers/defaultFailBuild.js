exports.defaultFailBuild = function (message, { error }) {
  throw new Error(`${message}\n${error.stack}`)
}
