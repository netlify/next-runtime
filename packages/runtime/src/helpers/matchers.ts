import { transform } from 'regexp-tree'

// The Go regexp lib doesn't support lookaheads, so we need to remove them
export const stripLookahead = (regex: string) => {
  // Early return if there's no lookahead
  if (!regex?.includes('(?!')) {
    return regex
  }
  try {
    // Parse the regexp into an AST
    const re = transform(`/${regex}/`, {
      Assertion(path) {
        // Remove the lookahead
        if (path.node.kind === 'Lookahead') {
          path.remove()
        }
      },
    })
    // Strip the leading and trailing slashes
    return re.toString().slice(1, -1)
  } catch {
    //  Failed to parse regex, so return unchanged
    return regex
  }
}

// The Go regexp lib has alternative syntax for named capture groups
export const transformCaptureGroups = (regex: string) => regex.replace(/\(\?<\w+>/, '(')

const LOCALIZED_REGEX_PREFIX_13_1 = '(?:\\/(_next\\/data\\/[^/]{1,}))?(?:\\/([^/.]{1,}))'
const OPTIONAL_REGEX_PREFIX_13_1 = '(?:\\/(_next\\/data\\/[^/]{1,}))?(?:\\/([^/.]{1,}))?'

const LOCALIZED_REGEX_PREFIX_13_3 = '(?:\\/(_next\\/data\\/[^/]{1,}))?(?:\\/((?!_next\\/)[^/.]{1,}))'
const OPTIONAL_REGEX_PREFIX_13_3 = '(?:\\/(_next\\/data\\/[^/]{1,}))?(?:\\/((?!_next\\/)[^/.]{1,}))?'

// Make the locale section of the matcher regex optional
export const makeLocaleOptional = (regex: string) => regex.replace(LOCALIZED_REGEX_PREFIX_13_1, OPTIONAL_REGEX_PREFIX_13_1).replace(LOCALIZED_REGEX_PREFIX_13_3, OPTIONAL_REGEX_PREFIX_13_3)
