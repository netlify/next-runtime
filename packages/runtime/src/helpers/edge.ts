import { promises as fs, existsSync } from 'fs'
import { resolve, join } from 'path'

import type { NetlifyConfig, NetlifyPluginConstants } from '@netlify/build'
import { greenBright } from 'chalk'
import destr from 'destr'
import { copy, copyFile, emptyDir, ensureDir, readJSON, writeJSON, writeJson } from 'fs-extra'
import type { PrerenderManifest } from 'next/dist/build'
import type { MiddlewareManifest } from 'next/dist/build/webpack/plugins/middleware-plugin'
import type { RouteHas } from 'next/dist/lib/load-custom-routes'
import { outdent } from 'outdent'

import { IMAGE_FUNCTION_NAME } from '../constants'

import { getRequiredServerFiles, NextConfig } from './config'
import { getPluginVersion } from './functionsMetaData'
import { makeLocaleOptional, stripLookahead, transformCaptureGroups } from './matchers'
import { RoutesManifest } from './types'
// This is the format as of next@12.2
interface EdgeFunctionDefinitionV1 {
  env: string[]
  files: string[]
  name: string
  page: string
  regexp: string
}

interface AssetRef {
  name: string
  filePath: string
}
export interface MiddlewareMatcher {
  regexp: string
  locale?: false
  has?: RouteHas[]
  missing?: RouteHas[]
}

// This is the format after next@12.3.0
interface EdgeFunctionDefinitionV2 {
  env: string[]
  files: string[]
  name: string
  page: string
  matchers: MiddlewareMatcher[]
  wasm?: AssetRef[]
  assets?: AssetRef[]
}

type EdgeFunctionDefinition = EdgeFunctionDefinitionV1 | EdgeFunctionDefinitionV2

export interface FunctionManifest {
  version: 1
  functions: Array<
    | {
        function: string
        name?: string
        path: string
        cache?: 'manual'
        generator: string
      }
    | {
        function: string
        name?: string
        pattern: string
        cache?: 'manual'
        generator: string
      }
  >
  layers?: Array<{ name: `https://${string}/mod.ts`; flag: string }>
  import_map?: string
}

const maybeLoadJson = <T>(path: string): Promise<T> | null => {
  if (existsSync(path)) {
    return readJSON(path)
  }
}
export const isAppDirRoute = (route: string, appPathRoutesManifest: Record<string, string> | null): boolean =>
  Boolean(appPathRoutesManifest) && Object.values(appPathRoutesManifest).includes(route)

export const loadMiddlewareManifest = (netlifyConfig: NetlifyConfig): Promise<MiddlewareManifest | null> =>
  maybeLoadJson(resolve(netlifyConfig.build.publish, 'server', 'middleware-manifest.json'))

export const loadAppPathRoutesManifest = (netlifyConfig: NetlifyConfig): Promise<Record<string, string> | null> =>
  maybeLoadJson(resolve(netlifyConfig.build.publish, 'app-path-routes-manifest.json'))

export const loadPrerenderManifest = (netlifyConfig: NetlifyConfig): Promise<PrerenderManifest> =>
  readJSON(resolve(netlifyConfig.build.publish, 'prerender-manifest.json'))

/**
 * Convert the Next middleware name into a valid Edge Function name
 */
const sanitizeName = (name: string) => `next_${name.replace(/\W/g, '_')}`

// Slightly different spacing in different versions!
const IMPORT_UNSUPPORTED = [
  `Object.defineProperty(globalThis,"__import_unsupported"`,
  `    Object.defineProperty(globalThis, "__import_unsupported"`,
]
/**
 * Concatenates the Next edge function code with the required chunks and adds an export
 */
const getMiddlewareBundle = async ({
  edgeFunctionDefinition,
  netlifyConfig,
}: {
  edgeFunctionDefinition: EdgeFunctionDefinition
  netlifyConfig: NetlifyConfig
}): Promise<string> => {
  const { publish } = netlifyConfig.build

  const shims = await fs.readFile(getEdgeTemplatePath('shims.js'), 'utf8')

  const chunks: Array<string> = [shims]

  chunks.push(`export const _DEFINITION = ${JSON.stringify(edgeFunctionDefinition)}`)

  if ('wasm' in edgeFunctionDefinition) {
    for (const { name, filePath } of edgeFunctionDefinition.wasm) {
      const wasm = await fs.readFile(join(publish, filePath))
      chunks.push(`const ${name} = _base64Decode(${JSON.stringify(wasm.toString('base64'))}).buffer`)
    }
  }

  if ('assets' in edgeFunctionDefinition) {
    chunks.push(`const _ASSETS = {}`)
    for (const { name, filePath } of edgeFunctionDefinition.assets) {
      const wasm = await fs.readFile(join(publish, filePath))
      chunks.push(`_ASSETS[${JSON.stringify(name)}] = ${JSON.stringify(wasm.toString('base64'))}`)
    }
  }

  for (const file of edgeFunctionDefinition.files) {
    const filePath = join(publish, file)

    let data = await fs.readFile(filePath, 'utf8')
    // Next defines an immutable global variable, which is fine unless you have more than one in the bundle
    // This adds a check to see if the global is already defined
    data = IMPORT_UNSUPPORTED.reduce(
      (acc, val) => acc.replace(val, `('__import_unsupported' in globalThis)||${val}`),
      data,
    )
    chunks.push('{', data, '}')
  }

  const exports = /* js */ `export default _ENTRIES["middleware_${edgeFunctionDefinition.name}"].default;`
  chunks.push(exports)
  return chunks.join('\n')
}

const getEdgeTemplatePath = (file: string) => join(__dirname, '..', '..', 'src', 'templates', 'edge', file)

const copyEdgeSourceFile = ({
  file,
  target,
  edgeFunctionDir,
}: {
  file: string
  edgeFunctionDir: string
  target?: string
}) => fs.copyFile(getEdgeTemplatePath(file), join(edgeFunctionDir, target ?? file))

const writeEdgeFunction = async ({
  edgeFunctionDefinition,
  edgeFunctionRoot,
  netlifyConfig,
  functionName,
  matchers = [],
  middleware = false,
}: {
  edgeFunctionDefinition: EdgeFunctionDefinition
  edgeFunctionRoot: string
  netlifyConfig: NetlifyConfig
  functionName: string
  matchers?: Array<MiddlewareMatcher>
  middleware?: boolean
}) => {
  const edgeFunctionDir = join(edgeFunctionRoot, functionName)

  const bundle = await getMiddlewareBundle({
    edgeFunctionDefinition,
    netlifyConfig,
  })

  await ensureDir(edgeFunctionDir)
  await fs.writeFile(join(edgeFunctionDir, 'bundle.js'), bundle)

  await copyEdgeSourceFile({
    edgeFunctionDir,
    file: middleware ? 'middleware-runtime.ts' : 'function-runtime.ts',
    target: 'index.ts',
  })

  if (middleware) {
    // Functions don't have complex matchers, so we can rely on the Netlify matcher
    await writeJson(join(edgeFunctionDir, 'matchers.json'), matchers)
  }
}

const generateEdgeFunctionMiddlewareMatchers = ({
  edgeFunctionDefinition,
  nextConfig,
}: {
  edgeFunctionDefinition: EdgeFunctionDefinition
  edgeFunctionRoot: string
  nextConfig: NextConfig
  cache?: 'manual'
}): Array<MiddlewareMatcher> => {
  // The v1 middleware manifest has a single regexp, but the v2 has an array of matchers
  if ('regexp' in edgeFunctionDefinition) {
    return [{ regexp: edgeFunctionDefinition.regexp }]
  }
  if (nextConfig.i18n) {
    return edgeFunctionDefinition.matchers.map((matcher) => ({
      ...matcher,
      regexp: makeLocaleOptional(matcher.regexp),
    }))
  }
  return edgeFunctionDefinition.matchers
}

const middlewareMatcherToEdgeFunctionDefinition = (
  matcher: MiddlewareMatcher,
  name: string,
  generator: string,
  cache?: 'manual',
): {
  function: string
  name?: string
  pattern: string
  cache?: 'manual'
  generator: string
} => {
  const pattern = transformCaptureGroups(stripLookahead(matcher.regexp))
  return { function: name, pattern, name, cache, generator }
}

export const cleanupEdgeFunctions = ({
  INTERNAL_EDGE_FUNCTIONS_SRC = '.netlify/edge-functions',
}: NetlifyPluginConstants) => emptyDir(INTERNAL_EDGE_FUNCTIONS_SRC)

export const writeDevEdgeFunction = async ({
  INTERNAL_EDGE_FUNCTIONS_SRC = '.netlify/edge-functions',
}: NetlifyPluginConstants) => {
  const generator = await getPluginVersion()
  const manifest: FunctionManifest = {
    functions: [
      {
        function: 'next-dev',
        name: 'netlify dev handler',
        path: '/*',
        generator,
      },
    ],
    version: 1,
  }
  const edgeFunctionRoot = resolve(INTERNAL_EDGE_FUNCTIONS_SRC)
  await emptyDir(edgeFunctionRoot)
  await writeJson(join(edgeFunctionRoot, 'manifest.json'), manifest)
  await copy(getEdgeTemplatePath('../edge-shared'), join(edgeFunctionRoot, 'edge-shared'))

  const edgeFunctionDir = join(edgeFunctionRoot, 'next-dev')
  await ensureDir(edgeFunctionDir)
  await copyEdgeSourceFile({ edgeFunctionDir, file: 'next-dev.js', target: 'index.js' })
}

/**
 * Writes an edge function that routes RSC data requests to the `.rsc` route
 */

export const generateRscDataEdgeManifest = async ({
  prerenderManifest,
  appPathRoutesManifest,
}: {
  prerenderManifest?: PrerenderManifest
  appPathRoutesManifest?: Record<string, string>
}): Promise<FunctionManifest['functions']> => {
  const generator = await getPluginVersion()
  if (!prerenderManifest || !appPathRoutesManifest) {
    return []
  }
  const staticAppdirRoutes: Array<string> = []
  for (const [path, route] of Object.entries(prerenderManifest.routes)) {
    if (isAppDirRoute(route.srcRoute, appPathRoutesManifest) && route.dataRoute) {
      staticAppdirRoutes.push(path, route.dataRoute)
    }
  }
  const dynamicAppDirRoutes: Array<string> = []

  for (const [path, route] of Object.entries(prerenderManifest.dynamicRoutes)) {
    if (isAppDirRoute(path, appPathRoutesManifest) && route.dataRouteRegex) {
      dynamicAppDirRoutes.push(route.routeRegex, route.dataRouteRegex)
    }
  }

  if (staticAppdirRoutes.length === 0 && dynamicAppDirRoutes.length === 0) {
    return []
  }

  const edgeFunctionDir = resolve('.netlify', 'edge-functions', 'rsc-data')
  await ensureDir(edgeFunctionDir)
  await copyEdgeSourceFile({ edgeFunctionDir, file: 'rsc-data.ts' })

  return [
    ...staticAppdirRoutes.map((path) => ({
      function: 'rsc-data',
      name: 'RSC data routing',
      path,
      generator,
    })),
    ...dynamicAppDirRoutes.map((pattern) => ({
      function: 'rsc-data',
      name: 'RSC data routing',
      pattern,
      generator,
    })),
  ]
}

export const getEdgeFunctionPatternForPage = ({
  edgeFunctionDefinition,
  pageRegexMap,
  appPathRoutesManifest,
}: {
  edgeFunctionDefinition: EdgeFunctionDefinition
  pageRegexMap: Map<string, string>
  appPathRoutesManifest?: Record<string, string>
}): string => {
  // We don't just use the matcher from the edge function definition, because it doesn't handle trailing slashes

  // appDir functions have a name that _isn't_ the route name, but rather the route with `/page` appended
  const regexp = pageRegexMap.get(appPathRoutesManifest?.[edgeFunctionDefinition.page] ?? edgeFunctionDefinition.page)
  if (regexp) {
    return regexp
  }
  if ('regexp' in edgeFunctionDefinition) {
    return edgeFunctionDefinition.regexp.replace(/([^/])\$$/, '$1/?$')
  }
  // If we need to fall back to the matcher, we need to add an optional trailing slash
  return edgeFunctionDefinition.matchers?.[0].regexp.replace(/([^/])\$$/, '$1/?$')
}

/**
 * Writes Edge Functions for the Next middleware
 */

// eslint-disable-next-line max-lines-per-function
export const writeEdgeFunctions = async ({
  netlifyConfig,
  routesManifest,
}: {
  netlifyConfig: NetlifyConfig
  routesManifest: RoutesManifest
}) => {
  const generator = await getPluginVersion()

  const manifest: FunctionManifest = {
    functions: [],
    layers: [],
    version: 1,
  }

  const edgeFunctionRoot = resolve('.netlify', 'edge-functions')
  await emptyDir(edgeFunctionRoot)

  const { publish } = netlifyConfig.build
  const nextConfigFile = await getRequiredServerFiles(publish)
  const nextConfig = nextConfigFile.config
  const usesAppDir = nextConfig.experimental?.appDir

  await copy(getEdgeTemplatePath('../edge-shared'), join(edgeFunctionRoot, 'edge-shared'))
  await writeJSON(join(edgeFunctionRoot, 'edge-shared', 'nextConfig.json'), nextConfig)
  await copy(join(publish, 'prerender-manifest.json'), join(edgeFunctionRoot, 'edge-shared', 'prerender-manifest.json'))

  // early return if edge is disabled
  if (destr(process.env.NEXT_DISABLE_NETLIFY_EDGE)) {
    console.log('Environment variable NEXT_DISABLE_NETLIFY_EDGE has been set, skipping Netlify Edge Function creation.')
    return
  }

  const rscFunctions = await generateRscDataEdgeManifest({
    prerenderManifest: await loadPrerenderManifest(netlifyConfig),
    appPathRoutesManifest: await loadAppPathRoutesManifest(netlifyConfig),
  })

  manifest.functions.push(...rscFunctions)

  const middlewareManifest = await loadMiddlewareManifest(netlifyConfig)
  if (!middlewareManifest) {
    console.error("Couldn't find the middleware manifest")
    return
  }

  let usesEdge = false

  for (const middleware of middlewareManifest.sortedMiddleware) {
    usesEdge = true
    const edgeFunctionDefinition = middlewareManifest.middleware[middleware]
    const functionName = sanitizeName(edgeFunctionDefinition.name)
    const matchers = generateEdgeFunctionMiddlewareMatchers({
      edgeFunctionDefinition,
      edgeFunctionRoot,
      nextConfig,
    })
    await writeEdgeFunction({
      edgeFunctionDefinition,
      edgeFunctionRoot,
      netlifyConfig,
      functionName,
      matchers,
      middleware: true,
    })

    manifest.functions.push(
      ...matchers.map((matcher) => middlewareMatcherToEdgeFunctionDefinition(matcher, functionName, generator)),
    )
  }
  // Functions (i.e. not middleware, but edge SSR and API routes)
  if (typeof middlewareManifest.functions === 'object') {
    // When using the app dir, we also need to check if the EF matches a page
    const appPathRoutesManifest = await loadAppPathRoutesManifest(netlifyConfig)

    // A map of all route pages to their page regex. This is used for pages dir and appDir.
    const pageRegexMap = new Map(
      [...(routesManifest.dynamicRoutes || []), ...(routesManifest.staticRoutes || [])].map((route) => [
        route.page,
        route.regex,
      ]),
    )
    // Create a map of pages-dir routes to their data route regex (appDir uses the same route as the HTML)
    const dataRoutesMap = new Map(
      [...(routesManifest.dataRoutes || [])].map((route) => [route.page, route.dataRouteRegex]),
    )

    for (const edgeFunctionDefinition of Object.values(middlewareManifest.functions)) {
      usesEdge = true
      const functionName = sanitizeName(edgeFunctionDefinition.name)
      await writeEdgeFunction({
        edgeFunctionDefinition,
        edgeFunctionRoot,
        netlifyConfig,
        functionName,
      })
      const pattern = getEdgeFunctionPatternForPage({
        edgeFunctionDefinition,
        pageRegexMap,
        appPathRoutesManifest,
      })
      manifest.functions.push({
        function: functionName,
        name: edgeFunctionDefinition.name,
        pattern,
        // cache: "manual" is currently experimental, so we restrict it to sites that use experimental appDir
        cache: usesAppDir ? 'manual' : undefined,
        generator,
      })
      // pages-dir page routes also have a data route. If there's a match, add an entry mapping that to the function too
      const dataRoute = dataRoutesMap.get(edgeFunctionDefinition.page)
      if (dataRoute) {
        manifest.functions.push({
          function: functionName,
          name: edgeFunctionDefinition.name,
          pattern: dataRoute,
          cache: usesAppDir ? 'manual' : undefined,
          generator,
        })
      }
    }
  }

  if (
    destr(process.env.NEXT_FORCE_EDGE_IMAGES) &&
    !destr(process.env.NEXT_DISABLE_EDGE_IMAGES) &&
    !destr(process.env.DISABLE_IPX)
  ) {
    usesEdge = true
    console.log(
      'Using Netlify Edge Functions for image format detection. Set env var "NEXT_DISABLE_EDGE_IMAGES=true" to disable.',
    )
    const edgeFunctionDir = join(edgeFunctionRoot, 'ipx')
    await ensureDir(edgeFunctionDir)
    await copyEdgeSourceFile({ edgeFunctionDir, file: 'ipx.ts', target: 'index.ts' })
    await copyFile(
      join('.netlify', 'functions-internal', IMAGE_FUNCTION_NAME, 'imageconfig.json'),
      join(edgeFunctionDir, 'imageconfig.json'),
    )
    manifest.functions.push({
      function: 'ipx',
      name: 'next/image handler',
      path: nextConfig.images.path || '/_next/image',
      generator,
    })

    manifest.layers.push({
      name: 'https://ipx-edge-function-layer.netlify.app/mod.ts',
      flag: 'ipx-edge-function-layer-url',
    })
  } else {
    console.log(
      'You are not using Netlify Edge Functions for image format detection. Set env var "NEXT_FORCE_EDGE_IMAGES=true" to enable.',
    )
  }

  if (usesEdge) {
    console.log(outdent`
      ✨ Deploying middleware and functions to ${greenBright`Netlify Edge Functions`} ✨
    `)
  }
  await writeJson(join(edgeFunctionRoot, 'manifest.json'), manifest)
}
