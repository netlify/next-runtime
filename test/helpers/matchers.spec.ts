import { getEdgeFunctionPatternForPage, sanitizeEdgePath } from '../../packages/runtime/src/helpers/edge'
import { makeLocaleOptional, stripLookahead } from '../../packages/runtime/src/helpers/matchers'

const makeDataPath = (path: string) => `/_next/data/build-id${path === '/' ? '/index' : path}.json`

const checkPath = (path: string, regex: string) => {
  const re = new RegExp(regex)
  const dataPath = makeDataPath(path)
  const testPath = re.test(path)
  const testData = re.test(dataPath)
  // For easier debugging
  // console.log({ path, regex, dataPath, testPath, testData })
  return testPath && testData
}

describe('the middleware path matcher', () => {
  it('makes the locale slug optional in the regex for the root', () => {
    // The regex generated by Next for the path "/" with i18n enabled
    const regex =
      '^(?:\\/(_next\\/data\\/[^/]{1,}))?(?:\\/([^/.]{1,}))(|\\.json|\\/?index|\\/?index\\.json)?[\\/#\\?]?$'
    expect(checkPath('/', regex)).toBe(false)
    expect(checkPath('/', makeLocaleOptional(regex))).toBe(true)
    expect(checkPath('/en', makeLocaleOptional(regex))).toBe(true)
  })

  it('makes the locale slug optional in the regex for a subpath', () => {
    // The regex generated by Next for the path "/static" with i18n enabled (<= 13.1.6)
    const regex = '^(?:\\/(_next\\/data\\/[^/]{1,}))?(?:\\/([^/.]{1,}))\\/static(.json)?[\\/#\\?]?$'
    // The regex generated by Next for the path "/static" with i18n enabled (>= 13.3.0)
    const regexTwo = '^(?:\\/(_next\\/data\\/[^/]{1,}))?(?:\\/((?!_next\\/)[^/.]{1,}))\\/static(.json)?[\\/#\\?]?$'

    const assertion = (input) => {
      expect(checkPath('/static', input)).toBe(false)
      expect(checkPath('/static', makeLocaleOptional(input))).toBe(true)
      expect(checkPath('/en/static', makeLocaleOptional(input))).toBe(true)
    }

    assertion(regex)
    assertion(regexTwo)
  })

  it('does not change the regex when calling makeLocaleOptional with a regex that has no locale', () => {
    const regexes = [
      '^(?:\\/(_next\\/data\\/[^/]{1,}))?(?:\\/(\\/?index|\\/?index\\.json))?[\\/#\\?]?$',
      '^(?:\\/(_next\\/data\\/[^/]{1,}))?\\/api(?:\\/((?:[^\\/#\\?]+?)(?:\\/(?:[^\\/#\\?]+?))*))?(.json)?[\\/#\\?]?$',
      '^(?:\\/(_next\\/data\\/[^/]{1,}))?\\/shows(?:\\/((?!99|88).*))(.json)?[\\/#\\?]?$',
    ]
    for (const regex of regexes) {
      expect(makeLocaleOptional(regex)).toBe(regex)
    }
  })

  it('removes lookaheads from the regex', () => {
    const regexes = [
      '^(?:\\/(_next\\/data\\/[^/]{1,}))?(?:\\/([^/.]{1,}))\\/shows(?:\\/((?!99|88).*))(.json)?[\\/#\\?]?$',
      '^(?:\\/(_next\\/data\\/[^/]{1,}))?\\/shows(?:\\/((?!99|88).*))(.json)?[\\/#\\?]?$',
    ]
    for (const regex of regexes) {
      const stripped = stripLookahead(regex)
      expect(regex).toMatch(/\(\?!/)
      expect(stripped).not.toMatch(/\(\?!/)
    }
  })
  it('converts regexes with lookaheads to stripped ones that still match at least the same paths', () => {
    const regex = '^(?:\\/(_next\\/data\\/[^/]{1,}))?\\/shows(?:\\/((?!99|88).*))(.json)?[\\/#\\?]?$'
    expect(checkPath('/shows', regex)).toBe(false)
    expect(checkPath('/shows/11', regex)).toBe(true)
    expect(checkPath('/shows/99', regex)).toBe(false)
    expect(checkPath('/shows/888', regex)).toBe(false)

    const stripped = stripLookahead(regex)
    expect(checkPath('/shows', stripped)).toBe(false)
    expect(checkPath('/shows/11', stripped)).toBe(true)
    // These will be true because the regex is not as strict as the original one
    // The strict test will be done in the JS entrypoint
    expect(checkPath('/shows/99', stripped)).toBe(true)
    expect(checkPath('/shows/888', stripped)).toBe(true)
  })
})

const pageRegexMap = new Map([
  ['/api/shows/[id]', '^/api/shows/([^/]+?)(?:/)?$'],
  ['/api/shows/[...params]', '^/api/shows/(.+?)(?:/)?$'],
  ['/app-edge/[id]', '^/app\\-edge/([^/]+?)(?:/)?$'],
  ['/blog/[author]', '^/blog/([^/]+?)(?:/)?$'],
  ['/blog/[author]/[slug]', '^/blog/([^/]+?)/([^/]+?)(?:/)?$'],
  ['/getServerSideProps/all/[[...slug]]', '^/getServerSideProps/all(?:/(.+?))?(?:/)?$'],
  ['/getServerSideProps/[id]', '^/getServerSideProps/([^/]+?)(?:/)?$'],
])

const appPathRoutesManifest = {
  '/app-edge/[id]/page': '/app-edge/[id]',
  '/blog/[author]/[slug]/page': '/blog/[author]/[slug]',
  '/blog/[author]/page': '/blog/[author]',
}

describe('the edge function matcher helpers', () => {
  it('finds the correct regex for an edge API route', () => {
    const regex = getEdgeFunctionPatternForPage({
      pageRegexMap,
      appPathRoutesManifest,
      edgeFunctionDefinition: {
        name: 'pages/api/og',
        page: '/api/og',
        env: [],
        files: [],
        matchers: [
          {
            regexp: '^/api/og$',
          },
        ],
        wasm: [],
        assets: [],
      },
    })
    expect(regex).toBe('^/api/og/?$')
    expect('/api/og').toMatch(new RegExp(regex))
    expect('/api/og/').toMatch(new RegExp(regex))
  })

  it('finds the correct regex for an appDir page with a dynamic route', () => {
    const regex = getEdgeFunctionPatternForPage({
      pageRegexMap,
      appPathRoutesManifest,
      edgeFunctionDefinition: {
        env: [],
        files: [],
        name: 'app/app-edge/[id]/page',
        page: '/app-edge/[id]/page',
        matchers: [
          {
            regexp: '^/app\\-edge/(?<id>[^/]+?)$',
          },
        ],
        wasm: [],
        assets: [],
      },
    })
    expect(regex).toBe('^/app\\-edge/([^/]+?)(?:/)?$')
    expect('/app-edge/1').toMatch(new RegExp(regex))
    expect('/app-edge/1/').toMatch(new RegExp(regex))
  })

  it('finds the correct regex for a pages edge route', () => {
    const regex = getEdgeFunctionPatternForPage({
      pageRegexMap,
      appPathRoutesManifest,
      edgeFunctionDefinition: {
        env: [],
        files: [],
        name: 'pages/edge/[id]',
        page: '/edge/[id]',
        matchers: [
          {
            regexp: '^/edge/(?<id>[^/]+?)$',
          },
        ],
        wasm: [],
        assets: [],
      },
    })
    expect(regex).toBe('^/edge/(?<id>[^/]+?)/?$')
    expect('/edge/1').toMatch(new RegExp(regex))
    expect('/edge/1/').toMatch(new RegExp(regex))
  })

  it('finds the correct regex for a pages edge route with a v1 definition', () => {
    const regex = getEdgeFunctionPatternForPage({
      pageRegexMap,
      appPathRoutesManifest,
      edgeFunctionDefinition: {
        env: [],
        files: [],
        name: 'pages/edge/[id]',
        page: '/edge/[id]',
        regexp: '^/edge/(?<id>[^/]+?)$',
        wasm: [],
        assets: [],
      },
    })
    expect(regex).toBe('^/edge/(?<id>[^/]+?)/?$')
    expect('/edge/1').toMatch(new RegExp(regex))
    expect('/edge/1/').toMatch(new RegExp(regex))
  })
})

describe('the images path is sanitized', () => {
  it('passes through a path', () => {
    expect(sanitizeEdgePath('/_next/image')).toBe('/_next/image')
  })

  it('strips domains', () => {
    expect(sanitizeEdgePath('http://example.com/_next/image')).toBe('/_next/image')
    expect(sanitizeEdgePath('https://example.com/_next/image')).toBe('/_next/image')
  })

  it('strips domains with globs', () => {
    expect(sanitizeEdgePath('http://example.com/_next/image/*')).toBe('/_next/image/*')
    expect(sanitizeEdgePath('https://example.com/_next/image/*')).toBe('/_next/image/*')
  })

  it('strips domains with tokens', () => {
    expect(sanitizeEdgePath('http://example.com/_next/image/:slug/')).toBe('/_next/image/:slug/')
    expect(sanitizeEdgePath('https://example.com/_next/image/:slug/')).toBe('/_next/image/:slug/')
  })
})
