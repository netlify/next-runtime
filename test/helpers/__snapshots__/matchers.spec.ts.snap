// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the middleware path matcher removes lookaheads from the regex 1`] = `"^(?:\\\\/(_next\\\\/data\\\\/[^/]{1,}))?(?:\\\\/([^/.]{1,}))\\\\/api(?:\\\\/((?:[^\\\\/#\\\\?]+?)(?:\\\\/(?:[^\\\\/#\\\\?]+?))*))?(.json)?[\\\\/#\\\\?]?$"`;

exports[`the middleware path matcher removes lookaheads from the regex 2`] = `"^(?:\\\\/(_next\\\\/data\\\\/[^/]{1,}))?(?:\\\\/([^/.]{1,}))\\\\/headers(.json)?[\\\\/#\\\\?]?$"`;

exports[`the middleware path matcher removes lookaheads from the regex 3`] = `"^(?:\\\\/(_next\\\\/data\\\\/[^/]{1,}))?(?:\\\\/([^/.]{1,}))\\\\/cookies(?:\\\\/((?:[^\\\\/#\\\\?]+?)(?:\\\\/(?:[^\\\\/#\\\\?]+?))*))?(.json)?[\\\\/#\\\\?]?$"`;

exports[`the middleware path matcher removes lookaheads from the regex 4`] = `"^(?:\\\\/(_next\\\\/data\\\\/[^/]{1,}))?(?:\\\\/([^/.]{1,}))\\\\/static(.json)?[\\\\/#\\\\?]?$"`;

exports[`the middleware path matcher removes lookaheads from the regex 5`] = `"^(?:\\\\/(_next\\\\/data\\\\/[^/]{1,}))?(?:\\\\/([^/.]{1,}))\\\\/matcher-cookie(.json)?[\\\\/#\\\\?]?$"`;

exports[`the middleware path matcher removes lookaheads from the regex 6`] = `"^(?:\\\\/(_next\\\\/data\\\\/[^/]{1,}))?(?:\\\\/([^/.]{1,}))\\\\/shows(?:\\\\/(.*))(.json)?[\\\\/#\\\\?]?$"`;
